src/metabase/lib/metadata.cljc

```diff
@@ -0,0 +1,166 @@
+(ns metabase.lib.metadata
+  (:require
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.lib.util :as lib.util]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(def FieldMetadata
+  "Malli schema for a valid map of Field metadata."
+  [:map
+   [:lib/type [:= :metadata/field]]
+   [:id {:optional true} ::lib.schema.id/field]
+   [:name ::lib.schema.common/non-blank-string]])
+
+(def ^:private TableMetadata
+  [:map
+   [:lib/type [:= :metadata/table]]
+   [:id ::lib.schema.id/table]
+   [:name ::lib.schema.common/non-blank-string]
+   [:schema ::lib.schema.common/non-blank-string]
+   [:fields [:sequential FieldMetadata]]])
+
+(def DatabaseMetadata
+  "Malli schema for the DatabaseMetadata as returned by `GET /api/database/:id/metadata` -- what should be available to
+  the frontend Query Builder."
+  [:map
+   [:lib/type [:= :metadata/database]]
+   [:id ::lib.schema.id/database]
+   [:tables [:sequential TableMetadata]]])
+
+(def StageMetadata
+  "Malli schema for the results metadata (`[:data :results_metadata]`) that is included in query results, and saved as
+  `result_metadata` for a Saved Question.
+
+  Note that queries currently actually come back with both `data.results_metadata` AND `data.cols`; it looks like the
+  Frontend actually *merges* these together -- see `applyMetadataDiff` in
+  `frontend/src/metabase/query_builder/selectors.js` -- but this is ridiculous. Let's try to merge anything missing in
+  `results_metadata` into `cols` going forward so things don't need to be manually merged in the future."
+  [:map
+   [:lib/type [:= :metadata/results]]
+   [:columns [:sequential FieldMetadata]]])
+
+(defmulti ^:private database-metadata*
+  {:arglists '([x])}
+  lib.dispatch/dispatch-value)
+
+(defmethod database-metadata* :mbql/query
+  [query]
+  (:lib/metadata query))
```

Comment:
So we need to keep that Database metadata around somewhere I think in order to be able to power stuff like a list of Tables you can join against or stuff like that. If you're building a new query from scratch, you'd start with DatabaseMetadata and query it to see what sorts of things you can add to the query e.g. a filter on a particular Field, and then add it

I'm not 100% sure how it should work in the end but in #28717 I'm playing around with having the top-level `:lib/metadata` always be DatabaseMetadata (or something compatible) and then having `:lib/stage-metadata` attached to each individual stage (sort of like `:source-metadata` in QP MBQL) but I'm not 100% sure about that. 

Either way even if we have source query metadata then we'd still want DatabaseMetadata around in case you wanted to add an extra Field or join or something